### ISSUE TITLE
Python Azure function doesn't build requirements.txt

### ISSUE THREAD

**Issue by reubano** [pulumi/pulumi-azure-native#2797]
### What happened?

I created a Python Azure function with an HTTP trigger. In Azure, if the function uses a `requirements.txt` file, the requirements do not build and the function doesn't work. However, the function works fine locally. Additionally, if I deploy via the Azure CLI `npx func azure functionapp publish <app-name>`, the function also works as expected.

#### Resources used

StorageAccount
BlobContainer
Blob
WebApp
Workspace

#### Azure function invocation log

```bash
Executing 'Functions.Test' (Reason='This function was programmatically called via the host APIs.', Id=6af37f67-f86c-408a-9aa7-c189bf3eedb8)

Result: Failure Exception: ModuleNotFoundError: No module named 'azure.identity'. Please check the requirements.txt file for the missing module. For more info, please refer the troubleshooting guide: https://aka.ms/functions-modulenotfound Stack: File "/azure-functions-host/workers/python/3.9/LINUX/X64/azure_functions_worker/dispatcher.py", line 380, in _handle__function_load_request func = loader.load_function( File "/azure-functions-host/workers/python/3.9/LINUX/X64/azure_functions_worker/utils/wrappers.py", line 48, in call raise extend_exception_message(e, message) File "/azure-functions-host/workers/python/3.9/LINUX/X64/azure_functions_worker/utils/wrappers.py", line 44, in call return func(*args, **kwargs) File "/azure-functions-host/workers/python/3.9/LINUX/X64/azure_functions_worker/loader.py", line 132, in load_function mod = importlib.import_module(fullmodname) File "/usr/local/lib/python3.9/importlib/__init__.py", line 127, in import_module return _bootstrap._gcd_import(name[level:], package, level) File "/home/site/wwwroot/Test/__init__.py", line 4, in <module> from azure.identity import DefaultAzureCredential 
```

### Example

```bash
$ tree -I 'venv|__pycache__' 
.
├── Pulumi.yaml
├── Test
│   ├── Test
│   │   ├── __init__.py
│   │   └── function.json
│   ├── host.json
│   ├── local.settings.json
│   └── requirements.txt
├── __main__.py
└── requirements.txt
```
`Pulumi.yaml`

```yml
name: app
runtime:
  name: python
  options:
    virtualenv: venv
description: A minimal Azure Native Python Pulumi program
config:
  location: centralus
  email: rcummings@nerevu.com
  subscriptionId: xxx
  resourceGroup:
    value:
      name: xxx
      tags:
        CostCenter: xxx
        Environment: xxx
  resourceName: xxx
```

`__main__.py`

```python
from pulumi import Output, Config, export, FileArchive, ResourceOptions

from pulumi_azure_native import storage
from pulumi_azure_native.storage import StorageAccount, BlobContainer, Blob, list_storage_account_sas, HttpProtocol, Permissions
from pulumi_azure_native.resources import ResourceGroup
from pulumi_azure_native.web import AppServicePlan, WebApp, SiteConfigArgs, NameValuePairArgs, CorsSettingsArgs
from pulumi_azure_native.insights import Component
from pulumi_azure_native.operationalinsights import Workspace, WorkspaceSkuArgs

config = Config()
subscription_id = config.require("subscriptionId")
resource_name = config.require("resourceName")
resource_group_config = config.require_object("resourceGroup")
resource_group_name = resource_group_config["name"]


def get_signed_url(storage, container, blob, group):
    sas = Output.all(storage.name, group.name).apply(
        lambda args: list_storage_account_sas(
            account_name=args[0],
            permissions=Permissions.R,
            resource_group_name=args[1],
            resource_types="o",
            protocols=HttpProtocol.HTTPS,
            services="b",
            shared_access_expiry_time="2030-01-01",
        )
    )

    args = [storage.name, container.name, blob.name, sas.account_sas_token]
    return Output.format("https://{0}.blob.core.windows.net/{1}/{2}?{3}", *args)


def get_app_setting(storage, container, blob, group):
    instrumentation_key = component.instrumentation_key

    return [
        NameValuePairArgs(name="AzureWebJobsStorage", value=storage_account_connection_string),
        NameValuePairArgs(name="FUNCTIONS_EXTENSION_VERSION", value="~4"),
        NameValuePairArgs(name="FUNCTIONS_WORKER_RUNTIME", value="python"),
        NameValuePairArgs(name="WEBSITE_CONTENTAZUREFILECONNECTIONSTRING", value=storage_account_connection_string),
        NameValuePairArgs(name="WEBSITE_CONTENTSHARE", value=web_app_resource_name.lower()),
        NameValuePairArgs(name="APPINSIGHTS_INSTRUMENTATIONKEY", value=instrumentation_key),
        NameValuePairArgs(name="WEBSITE_RUN_FROM_PACKAGE", value=get_signed_url(storage, container, blob, group)),
        NameValuePairArgs(name="ENABLE_ORYX_BUILD", value=True),
        NameValuePairArgs(name="SCM_DO_BUILD_DURING_DEPLOYMENT", value=True),
    ]

resource_group = ResourceGroup(
    resource_name,
    location="centralus",
    resource_group_name=resource_group_name,
    tags=resource_group_config["tags"],
    opts=ResourceOptions(protect=True),
)

storage_account = StorageAccount(
    "BugFixStorage",
    minimum_tls_version="TLS1_2",
    account_name="bugfixstorageacct",
    allow_blob_public_access=False,
    resource_group_name=resource_group.name,
    sku=storage.SkuArgs(name=storage.SkuName.STANDARD_LRS),
    kind=storage.Kind.STORAGE_V2,
)

primary_storage_account_key = Output.secret(
    Output.all(resource_group.name, storage_account.name).apply(
        lambda args: storage.list_storage_account_keys(
            resource_group_name=args[0], account_name=args[1]
        )
    ).apply(lambda keys: keys.keys[0].value)
)

storage_account_connection_string = Output.format(
    "DefaultEndpointsProtocol=https;AccountName={0};AccountKey={1}",
    storage_account.name,
    primary_storage_account_key,
)

workspace = Workspace(
    f"BugFixWorkspace",
    workspace_name=f"bugfix-workspace",
    location=resource_group.location,
    resource_group_name=resource_group.name,
    retention_in_days=30,
    sku=WorkspaceSkuArgs(name="PerGB2018"),
)

component = Component(
    f"BugFixComponent",
    resource_name_=f"bugfix-component",
    application_type="web",
    flow_type="Bluefield",
    kind="web",
    location=resource_group.location,
    request_source="rest",
    resource_group_name=resource_group.name,
    workspace_resource_id=workspace.id
)

storage_account_connection_string = Output.format(
    "DefaultEndpointsProtocol=https;AccountName={0};AccountKey={1}",
    storage_account.name,
    primary_storage_account_key
)
service_plan = AppServicePlan(
    "BugFixServicePlan",
    name="bugfix-service-plan",
    kind="Linux",
    reserved=True,
    location=resource_group.location,
    resource_group_name=resource_group.name,
    sku={"tier": "Dynamic", "name": "Y1"},
)

code_container = BlobContainer(
    "BugFixCodeContainer",
    container_name="bugfix-code-container",
    account_name=storage_account.name,
    resource_group_name=resource_group.name,
)

app_name = "bugfix-function"
func_name = "Test"
web_app_resource_name = "BugFixFunction"

code_blob = Blob(
    f"{func_name}Code",
    blob_name=f"{app_name}-code.zip",
    container_name=code_container.name,
    account_name=storage_account.name,
    resource_group_name=resource_group.name,
    source=FileArchive(func_name),
)

web_app = WebApp(
    web_app_resource_name,
    name=app_name,
    kind="functionapp",
    https_only=True,
    reserved=True,
    location=resource_group.location,
    resource_group_name=resource_group.name,
    server_farm_id=service_plan.id,
    identity={"type": "SystemAssigned"},
    site_config=SiteConfigArgs(
        python_version="3.9",
        ftps_state="Disabled",
        http20_enabled=True,
        linux_fx_version="PYTHON|3.9",
        cors=CorsSettingsArgs(
            allowed_origins=["https://portal.azure.com"],
        ),
        app_settings=get_app_setting(storage_account, code_container, code_blob, resource_group),
    ),
)

subscription_path = f"/subscriptions/{subscription_id}/resourceGroups/{resource_group_name}"
import_command = f"pulumi import azure-native:resources:ResourceGroup {resource_name} {subscription_path}"
export("loggingMessage", f"Run the following command to import the resource group first: {import_command}")

func_endpoint = web_app.default_host_name.apply(lambda name: "https://{0}/api/{1}".format(name, func_name))
export("FunctionEndpoint", func_endpoint)
```

`__init__.py`

```python
import logging
import azure.functions as func

from azure.identity import DefaultAzureCredential


def main(req: func.HttpRequest) -> func.HttpResponse:
    logging.info("Processing HTTP trigger function request…")

    if credentials := DefaultAzureCredential():
        response = func.HttpResponse(str(credentials), status_code=200)
    else:
        response = func.HttpResponse("No credentials!", status_code=401)

    return response
```

`function.json`

```json
{
  "scriptFile": "__init__.py",
  "bindings": [
    {
      "authLevel": "anonymous",
      "type": "httpTrigger",
      "direction": "in",
      "name": "req",
      "methods": [
        "get"
      ]
    },
    {
      "type": "http",
      "direction": "out",
      "name": "$return"
    }
  ]
}
```

`host.json`

```json
{
  "version": "2.0",
  "logging": {
    "applicationInsights": {
      "samplingSettings": {
        "isEnabled": true
      }
    }
  },
  "extensionBundle": {
    "id": "Microsoft.Azure.Functions.ExtensionBundle",
    "version": "[3.3.0, 4.0.0)"
  }
}
```

`local.settings.json`

```json
{
  "IsEncrypted": false,
  "Values": {
    "FUNCTIONS_WORKER_RUNTIME": "python",
    "FUNCTIONS_EXTENSION_VERSION": "~4",
    "LINUX_FX_VERSION": "Python|3.9",
    "linuxFxVersion": "Python|3.9",
  },
  "ConnectionStrings": {}
}
```

`Test/requirements.txt`

```
# DO NOT include azure-functions-worker in this file
# The Python Worker is managed by Azure Functions platform
# Manually managing azure-functions-worker may cause unexpected issues

azure-functions==1.11.2
azure-identity==1.14.0
```

`requirements.txt`

```
pulumi>=3.73.0,<4.0.0
pulumi-azure-native>=2.9.0,<3.0.0
```

### Output of `pulumi about`

```bash
CLI          
Version      3.86.0
Go Version   go1.21.1
Go Compiler  gc

Plugins
NAME          VERSION
azure-native  2.9.0
python        unknown

Host     
OS       darwin
Version  12.6.1
Arch     arm64

This project is written in python: executable='.../venv/bin/python3' version='3.11.5'

Current Stack: nerevu/app/dev

TYPE                                        URN
pulumi:pulumi:Stack                         urn:pulumi:dev::app::pulumi:pulumi:Stack::app-dev
pulumi:providers:azure-native               urn:pulumi:dev::app::pulumi:providers:azure-native::default_2_9_0
azure-native:resources:ResourceGroup        urn:pulumi:dev::app::azure-native:resources:ResourceGroup::xxx
azure-native:web:AppServicePlan             urn:pulumi:dev::app::azure-native:web:AppServicePlan::BugFixServicePlan
azure-native:storage:StorageAccount         urn:pulumi:dev::app::azure-native:storage:StorageAccount::BugFixStorage
pulumi:providers:azure-native               urn:pulumi:dev::app::pulumi:providers:azure-native::default
azure-native:storage:BlobContainer          urn:pulumi:dev::app::azure-native:storage:BlobContainer::BugFixCodeContainer
azure-native:storage:Blob                   urn:pulumi:dev::app::azure-native:storage:Blob::TestCode
azure-native:operationalinsights:Workspace  urn:pulumi:dev::app::azure-native:operationalinsights:Workspace::BugFixWorkspace
azure-native:insights:Component             urn:pulumi:dev::app::azure-native:insights:Component::BugFixComponent
azure-native:web:WebApp                     urn:pulumi:dev::app::azure-native:web:WebApp::BugFixFunction


Found no pending operations associated with nerevu/dev

Backend        
Name           pulumi.com
URL            https://app.pulumi.com/reubano
User           reubano
Organizations  reubano, nerevu
Token type     personal

Dependencies:
NAME                 VERSION
pip                  23.2.1
pulumi-azure-native  2.9.0
setuptools           65.5.0

Pulumi locates its logs in /var/folders/m1/z7tt4jpd23z27g6n21yyrh2h0000gp/T/ by default

```

### Additional context

`commands`

```bash
$ python -m venv venv
$ source venv/bin/activate
$ pip install -r requirements.txt
$ pulumi stack
$ pulumi import azure-native:resources:ResourceGroup <resourceName> /subscriptions/<subscriptionId>/resourceGroups/<resourceGroupName>
$ pulumi up
curl -v `pulumi stack output FunctionEndpoint`
...
< HTTP/2 500
...

$ npx func azure functionapp publish <app-name>
...
Remote build succeeded!
...
$ curl -v https://xxx.azurewebsites.net/api/test
< HTTP/2 200
...
```

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by mikhailshilkov** [comment]
Hi @reubano thank you for opening this issue.

> In Azure, if the function uses a requirements.txt file, the requirements do not build and the function doesn't work

This sounds like an Azure question, not a Pulumi question to me. Have you verified that the WebApp resource is deployed correctly and that it points to all the right files? If so, I think the provider works as intended and you need to figure out how to configure the Azure Function App to build the Python app.

**Comment by reubano** [comment]
Thanks @mikhailshilkov, that led me down the right path. I tested by deploying via the Azure CLI `npx func azure functionapp publish <app-name>` and it complained about a missing `local.settings.json`. After adding a bare-bones `local.settings.json`, the CLI publish command worked. I then redeployed via `pulumi up` and it worked as well.

I think it would be good for pulumi to detect when `local.settings.json` is missing and warn the user that they should add it.
